File: AlbumLibrary/src/main/java/com/album/util/scanner/SingleMediaScanner.java
package com.album.util.scanner;

import android.content.Context;
import android.media.MediaScannerConnection;
import android.net.Uri;

import com.album.ui.annotation.AlbumResultType;

import java.io.File;

/**
 * by y on 14/08/2017.
 * <p>
 * <p>
 * https://issuetracker.google.com/issues/37046656
 * <p>
 * https://github.com/square/leakcanary/issues/26
 */

public class SingleMediaScanner implements MediaScannerConnection.MediaScannerConnectionClient {

    private MediaScannerConnection connection;
    private final File file;
    private final SingleScannerListener listener;
    private final int type;

    public SingleMediaScanner(Context context, File file, SingleScannerListener listener, @AlbumResultType int type) {
        this.file = file;
        this.type = type;
        this.connection = new MediaScannerConnection(context.getApplicationContext(), this);
        this.connection.connect();
        this.listener = listener;
        this.listener.onScanStart();
    }

    @Override
    public void onMediaScannerConnected() {
        connection.scanFile(file.getAbsolutePath(), null);
    }

    public void disconnect() {
        if (connection != null) {
            connection.disconnect();
            connection = null;
        }
    }

    @Override
    public void onScanCompleted(String path, Uri uri) {
        disconnect();
        if (listener != null) {
            listener.onScanCompleted(type);
        }
    }
}

File: scan/src/main/java/com/gallery/scan/SingleMediaScanner.kt
package com.gallery.scan

import android.content.Context
import android.media.MediaScannerConnection
import android.net.Uri

class SingleMediaScanner(context: Context, private val path: String, private val type: Int, private val listener: SingleScannerListener) : MediaScannerConnection.MediaScannerConnectionClient {

    interface SingleScannerListener {
        fun onScanStart()

        fun onScanCompleted(type: Int, path: String?, uri: Uri?)
    }

    private val connection: MediaScannerConnection = MediaScannerConnection(context.applicationContext, this)

    init {
        connection.connect()
        listener.onScanStart()
    }

    override fun onMediaScannerConnected() {
        connection.scanFile(path, null)
    }

    fun disconnect() {
        connection.disconnect()
    }

    override fun onScanCompleted(path: String?, uri: Uri?) {
        listener.onScanCompleted(type, path, uri)
    }
}